pandas/_libs/tslibs/period.pyx

```diff
@@ -1722,8 +1716,15 @@ cdef class PeriodMixin:
             condition = self.freq != other
 
         if condition:
-            freqstr = freq_to_period_freqstr(self.freq.n, self.freq.name)
-            other_freqstr = freq_to_period_freqstr(other.n, other.name)
+            freqstr = PeriodDtypeBase(
+                self.freq._period_dtype_code, self.freq.n
+            )._freqstr
+            if hasattr(other, "_period_dtype_code"):
+                other_freqstr = PeriodDtypeBase(
+                    other._period_dtype_code, other.n
+                )._freqstr
+            else:
+                other_freqstr = other.freqstr
```

Comment:
why do we need the if-else branch for `other` but not for `self.freq`?

Reply:
I think because `self.freq` always has attribute `"_period_dtype_code",` bur `other` sometimes doesn't.
For example in many tests in `pandas/tests/arithmetic/test_period.py`( e.g. `test_pi_add_sub_timedeltalike_freq_mismatch_monthly`) other is equal `pd.offsets.YearBegin(2)`